src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,164 @@
+
+        cumulativeEigenLayerTokeEarnings[payment.token] += sumAmounts * eigenLayerShareBIPs / MAX_BIPS;
+        emit PaymentReceived(msg.sender, payment);
+    }
+
+    // @notice Permissioned function which allows posting a new Merkle root
+    function postMerkleRoot(bytes32 newRoot, uint256 height, uint256 calculatedUpToBlockNumber) external onlyRootPublisher{
+        MerkleRootPost memory newMerkleRoot = MerkleRootPost(newRoot, height, block.number + merkleRootActivationDelay, calculatedUpToBlockNumber);
+        merkleRoots.push(newMerkleRoot);
+        emit NewMerkleRootPosted(newMerkleRoot);
+    }
+
+    /// @notice Permissioned function which allows rootPublisher to nullify a Merkle root
+    function nullifyMerkleRoot(uint256 rootIndex) external onlyRootPublisher{
+        require(block.number <= merkleRoots[rootIndex].confirmedAtBlockNumber, "PaymentCoordinator.nullifyMerkleRoot: Merkle root already confirmed");
+        merkleRoots[rootIndex].root = bytes32(0);
+    }
+
+    // @notice Permissioned function which allows withdrawal of EigenLayer's share of `token` from all received payments
+    function withdrawEigenlayerShare(IERC20 token, address recipient) external onlyRootPublisher{
+        uint256 amount = cumulativeEigenLayerTokeEarnings[token];
+        token.safeTransfer(recipient, amount);
+        cumulativeEigenLayerTokeEarnings[token] = 0;
+    }
+
+    /**
+    * @notice Called by a staker or operator to prove the inclusion of their earnings in a posted Merkle root and claim them.
+    * @param proof Merkle proof showing that a leaf containing `(msg.sender, amount)` was included in the `rootIndex`-th
+    * Merkle root posted for the `token`
+    * @param rootIndex Specifies the node inside the Merkle tree corresponding to the specified root, `merkleRoots[rootIndex].root`.
+    */
+    function proveAndClaimEarnings(
+        bytes memory proof,
+        uint256 rootIndex,
+        MerkleLeaf memory leaf
+    ) external{
+        require(leaf.amounts.length == leaf.tokens.length, "PaymentCoordinator.proveAndClaimEarnings: leaf amounts and tokens must be same length");
+        require(merkleRoots[rootIndex].confirmedAtBlockNumber < block.number, "PaymentCoordinator.proveAndClaimEarnings: Merkle root not yet confirmed");
+        bytes32 root = merkleRoots[rootIndex].root;
+        require(root != bytes32(0), "PaymentCoordinator.proveAndClaimEarnings: Merkle root is null");
+        bytes32 leafHash = keccak256(abi.encodePacked(leaf.recipient, keccak256(abi.encodePacked(leaf.tokens)), keccak256(abi.encodePacked(leaf.amounts)), leaf.index));
+        require(Merkle.verifyInclusionKeccak(proof, root, leafHash, leaf.index), "PaymentCoordinator.proveAndClaimEarnings: Invalid proof");
+
+        for(uint i = 0; i < leaf.amounts.length; i++) {
+            leaf.tokens[i].safeTransfer(leaf.recipient, leaf.amounts[i]);
+            cumulativeTokenAmountClaimedByRecipient[leaf.tokens[i]][leaf.recipient] += leaf.amounts[i];
```

Comment:
this is incorrect, it should:

1. calculate the amount to send as (`leaf.amounts[i] - cumulativeTokenAmountClaimedByRecipient[leaf.tokens[i]][leaf.recipient]`)
2. set `cumulativeTokenAmountClaimedByRecipient[leaf.tokens[i]][leaf.recipient] = leaf.amounts[i]`
3. transfer the tokens

This is also not observing checks-effects-interactions ordering -- transfer after changing storage unless it's not possible.

Reply:
ok so leaf amounts represents cumulativeAmount + newAmount?  I guess I've misunderstood the interface

Reply:
> ok so leaf amounts represents cumulativeAmount + newAmount? I guess I've misunderstood the interface

leaf amounts should only ever be greater for new trees. I would describe the amounts as "(all time) cumulative earnings". formatting things like this means users can wait and do single proof later, rather than being forced to prove against every root

Reply:
done